File: backend/src/modules/tariff/fx.service.ts
import { Injectable, Logger, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, LessThanOrEqual } from 'typeorm';
import { FxRate } from './entities/fx-rate.entity';

@Injectable()
export class FxService {
  private readonly logger = new Logger(FxService.name);

  constructor(
    @InjectRepository(FxRate)
    private readonly fxRateRepository: Repository<FxRate>,
  ) {}

  async getRate(fromCcy: string, toCcy: string, date: Date): Promise<number> {
    const normalizedFromCcy = fromCcy.trim().toUpperCase();
    const normalizedToCcy = toCcy.trim().toUpperCase();

    if (normalizedFromCcy === normalizedToCcy) {
      return 1.0;
    }

    this.logger.debug(
      `Getting FX rate: ${normalizedFromCcy} -> ${normalizedToCcy} on ${date.toISOString()}`,
    );

    try {
      const directRate = await this.findDirectRate(normalizedFromCcy, normalizedToCcy, date);
      if (directRate !== null) {
        this.logger.debug(
          `Found direct rate: ${normalizedFromCcy}/${normalizedToCcy} = ${directRate}`,
        );
        return directRate;
      }

      const inverseRate = await this.findInverseRate(normalizedFromCcy, normalizedToCcy, date);
      if (inverseRate !== null) {
        this.logger.debug(
          `Found inverse rate: ${normalizedToCcy}/${normalizedFromCcy} = ${inverseRate}, returning ${1.0 / inverseRate}`,
        );
        return 1.0 / inverseRate;
      }

      throw new NotFoundException(
        `No FX rate found for ${normalizedFromCcy}/${normalizedToCcy} on or before ${date.toISOString().split('T')[0]}`,
      );
    } catch (error) {
      if (error instanceof NotFoundException) {
        throw error;
      }

      this.logger.error(
        `Error getting FX rate ${normalizedFromCcy}/${normalizedToCcy}: ${(error as Error).message}`,
        (error as Error).stack,
      );
      throw new Error(`FX rate lookup failed: ${(error as Error).message}`);
    }
  }

  private async findDirectRate(fromCcy: string, toCcy: string, date: Date): Promise<number | null> {
    try {
      const fxRate = await this.fxRateRepository.findOne({
        where: {
          from_ccy: fromCcy,
          to_ccy: toCcy,
          rate_date: LessThanOrEqual(date),
        },
        order: {
          rate_date: 'DESC',
        },
      });

      return fxRate ? Number(fxRate.rate) : null;
    } catch (error) {
      this.logger.warn(
        `Error in direct rate lookup ${fromCcy}/${toCcy}: ${(error as Error).message}`,
      );
      return null;
    }
  }

  private async findInverseRate(fromCcy: string, toCcy: string, date: Date): Promise<number | null> {
    try {
      const fxRate = await this.fxRateRepository.findOne({
        where: {
          from_ccy: toCcy,
          to_ccy: fromCcy,
          rate_date: LessThanOrEqual(date),
        },
        order: {
          rate_date: 'DESC',
        },
      });

      return fxRate ? Number(fxRate.rate) : null;
    } catch (error) {
      this.logger.warn(
        `Error in inverse rate lookup ${toCcy}/${fromCcy}: ${(error as Error).message}`,
      );
      return null;
    }
  }

  async bulkGetRates(
    requests: Array<{
      fromCcy: string;
      toCcy: string;
      date: Date;
    }>,
  ): Promise<Map<string, number>> {
    const results = new Map<string, number>();
    const cache = new Map<string, number>();

    for (const request of requests) {
      const cacheKey = `${request.fromCcy}-${request.toCcy}-${request.date.toDateString()}`;
      
      if (cache.has(cacheKey)) {
        results.set(cacheKey, cache.get(cacheKey)!);
        continue;
      }

      try {
        const rate = await this.getRate(request.fromCcy, request.toCcy, request.date);
        results.set(cacheKey, rate);
        cache.set(cacheKey, rate);
      } catch (error) {
        this.logger.warn(
          `Failed to get FX rate for ${cacheKey}: ${(error as Error).message}`,
        );
        // Continue processing other requests even if one fails
      }
    }

    return results;
  }

  async seedCommonRates(source: string): Promise<void> {
    const seedDate = new Date('2023-01-01');
    const rates = [
      { fromCcy: 'EUR', toCcy: 'CHF', rate: 0.9850 },
      { fromCcy: 'EUR', toCcy: 'USD', rate: 1.0650 },
      { fromCcy: 'EUR', toCcy: 'GBP', rate: 0.8850 },
      { fromCcy: 'EUR', toCcy: 'PLN', rate: 4.6800 },
    ];

    this.logger.log(`Seeding common FX rates with source: ${source}`);

    try {
      for (const rateData of rates) {
        const existingRate = await this.fxRateRepository.findOne({
          where: {
            rate_date: seedDate,
            from_ccy: rateData.fromCcy,
            to_ccy: rateData.toCcy,
          },
        });

        if (!existingRate) {
          const fxRate = this.fxRateRepository.create({
            rate_date: seedDate,
            from_ccy: rateData.fromCcy,
            to_ccy: rateData.toCcy,
            rate: rateData.rate,
            source: source,
          });

          await this.fxRateRepository.save(fxRate);
          
          this.logger.debug(
            `Seeded rate: ${rateData.fromCcy}/${rateData.toCcy} = ${rateData.rate}`,
          );
        } else {
          this.logger.debug(
            `Rate already exists: ${rateData.fromCcy}/${rateData.toCcy} on ${seedDate.toISOString().split('T')[0]}`,
          );
        }
      }

      this.logger.log(`Successfully seeded ${rates.length} FX rates`);
    } catch (error) {
      this.logger.error(
        `Error seeding common FX rates: ${(error as Error).message}`,
        (error as Error).stack,
      );
      throw error;
    }
  }

  async addRate(
    fromCcy: string,
    toCcy: string,
    rate: number,
    date: Date,
    source: string,
  ): Promise<void> {
    const normalizedFromCcy = fromCcy.trim().toUpperCase();
    const normalizedToCcy = toCcy.trim().toUpperCase();

    if (normalizedFromCcy === normalizedToCcy) {
      throw new Error('Cannot add FX rate for same currency pair');
    }

    if (rate <= 0) {
      throw new Error('FX rate must be positive');
    }

    try {
      const fxRate = this.fxRateRepository.create({
        rate_date: date,
        from_ccy: normalizedFromCcy,
        to_ccy: normalizedToCcy,
        rate: rate,
        source: source,
      });

      await this.fxRateRepository.save(fxRate);
      
      this.logger.log(
        `Added FX rate: ${normalizedFromCcy}/${normalizedToCcy} = ${rate} on ${date.toISOString().split('T')[0]} (${source})`,
      );
    } catch (error) {
      this.logger.error(
        `Error adding FX rate: ${(error as Error).message}`,
        (error as Error).stack,
      );
      throw error;
    }
  }

  async getAvailableCurrencies(date?: Date): Promise<string[]> {
    try {
      const query = this.fxRateRepository
        .createQueryBuilder('fx')
        .select('DISTINCT fx.from_ccy', 'currency')
        .union(
          this.fxRateRepository
            .createQueryBuilder('fx')
            .select('DISTINCT fx.to_ccy', 'currency'),
        );

      if (date) {
        query.where('fx.rate_date <= :date', { date });
      }

      const result = await query.orderBy('currency').getRawMany();
      const currencies = result.map(row => row.currency).filter(Boolean);
      
      // Add EUR as base currency if not present
      if (!currencies.includes('EUR')) {
        currencies.unshift('EUR');
      }

      return currencies.sort();
    } catch (error) {
      this.logger.error(
        `Error getting available currencies: ${(error as Error).message}`,
        (error as Error).stack,
      );
      return ['EUR']; // Fallback to EUR only
    }
  }
}

File: backend/app/services/fx_service.py
"""FX Currency Conversion Service — historical exchange rate lookups.

Port of backend_legacy/src/modules/tariff/fx.service.ts
Issue: #46

Key features:
- Direct rate lookup with inverse rate fallback
- functools.lru_cache for same-day lookups (performance)
- SQLAlchemy async query on fx_rate table (no RLS — global reference data)
- round_monetary() for all conversions
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from decimal import Decimal
from typing import Any

import structlog
from fastapi import HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.database import FxRate
from app.utils.round import round_monetary

logger = structlog.get_logger(__name__)


# ---------------------------------------------------------------------------
# Result types
# ---------------------------------------------------------------------------


@dataclass
class FxRateResult:
    """Result of a single FX rate lookup."""

    from_ccy: str
    to_ccy: str
    rate: Decimal
    rate_date: date
    method: str  # "direct", "inverse", or "same_currency"

    def to_dict(self) -> dict[str, Any]:
        return {
            "from_ccy": self.from_ccy,
            "to_ccy": self.to_ccy,
            "rate": str(self.rate),
            "rate_date": self.rate_date.isoformat(),
            "method": self.method,
        }


# ---------------------------------------------------------------------------
# Service
# ---------------------------------------------------------------------------


class FxService:
    """Looks up historical FX rates and converts monetary amounts.

    Queries the fx_rate table (no RLS — global reference data).
    Tries direct rate first, falls back to inverse rate (1/rate).

    Example usage:
        svc = FxService()
        rate = await svc.get_rate(db, "EUR", "CHF", date(2024, 3, 1))
        converted = await svc.convert(db, amount, "EUR", "CHF", date(2024, 3, 1))
    """

    def __init__(self) -> None:
        self.logger = structlog.get_logger(__name__)

    async def get_rate(
        self,
        db: AsyncSession,
        from_ccy: str,
        to_ccy: str,
        rate_date: date,
    ) -> Decimal:
        """Return the exchange rate from_ccy → to_ccy on or before rate_date.

        Tries direct rate first; falls back to 1/inverse_rate.
        Returns Decimal(1) for same-currency pairs.

        Args:
            db: Async DB session (fx_rate has no RLS).
            from_ccy: Source currency code (case-insensitive).
            to_ccy: Target currency code (case-insensitive).
            rate_date: Date for which the most recent rate is wanted.

        Returns:
            Exchange rate as Decimal, rounded to 8 decimal places.

        Raises:
            HTTPException(404): No rate found for the pair on or before rate_date.
            HTTPException(500): Unexpected DB error.
        """
        normalized_from = from_ccy.strip().upper()
        normalized_to = to_ccy.strip().upper()

        if normalized_from == normalized_to:
            return Decimal("1")

        self.logger.debug(
            "fx_rate_lookup_started",
            from_ccy=normalized_from,
            to_ccy=normalized_to,
            rate_date=rate_date.isoformat(),
        )

        try:
            direct = await self._find_direct_rate(db, normalized_from, normalized_to, rate_date)
            if direct is not None:
                self.logger.debug(
                    "fx_rate_found_direct",
                    from_ccy=normalized_from,
                    to_ccy=normalized_to,
                    rate=str(direct),
                )
                return direct

            inverse = await self._find_inverse_rate(db, normalized_from, normalized_to, rate_date)
            if inverse is not None:
                result = round_monetary(Decimal("1") / inverse, places=8)
                self.logger.debug(
                    "fx_rate_found_inverse",
                    from_ccy=normalized_from,
                    to_ccy=normalized_to,
                    inverse_rate=str(inverse),
                    computed_rate=str(result),
                )
                return result

        except HTTPException:
            raise
        except Exception as exc:
            self.logger.error(
                "fx_rate_lookup_failed",
                from_ccy=normalized_from,
                to_ccy=normalized_to,
                error=str(exc),
            )
            raise HTTPException(
                status_code=500,
                detail=f"FX rate lookup failed: {exc}",
            ) from exc

        raise HTTPException(
            status_code=404,
            detail=(
                f"No FX rate found for {normalized_from}/{normalized_to} "
                f"on or before {rate_date.isoformat()}"
            ),
        )

    async def convert(
        self,
        db: AsyncSession,
        amount: Decimal,
        from_ccy: str,
        to_ccy: str,
        rate_date: date,
    ) -> Decimal:
        """Convert amount from_ccy → to_ccy using the most recent rate.

        Args:
            db: Async DB session.
            amount: Amount to convert.
            from_ccy: Source currency.
            to_ccy: Target currency.
            rate_date: Date for which the most recent rate is wanted.

        Returns:
            Converted amount rounded via round_monetary().
        """
        rate = await self.get_rate(db, from_ccy, to_ccy, rate_date)
        return round_monetary(amount * rate)

    async def bulk_get_rates(
        self,
        db: AsyncSession,
        requests: list[dict[str, Any]],
    ) -> dict[str, Decimal]:
        """Fetch multiple rates with in-memory caching for duplicate keys.

        Each request dict must have keys: from_ccy, to_ccy, rate_date (date).
        Returns a mapping of cache_key → rate; failed lookups are skipped.

        Cache key format: "{FROM_CCY}-{TO_CCY}-{YYYY-MM-DD}"
        """
        results: dict[str, Decimal] = {}
        cache: dict[str, Decimal] = {}

        for req in requests:
            from_ccy: str = req["from_ccy"]
            to_ccy: str = req["to_ccy"]
            req_date: date = req["rate_date"]
            cache_key = f"{from_ccy.upper()}-{to_ccy.upper()}-{req_date.isoformat()}"

            if cache_key in cache:
                results[cache_key] = cache[cache_key]
                continue

            try:
                rate = await self.get_rate(db, from_ccy, to_ccy, req_date)
                results[cache_key] = rate
                cache[cache_key] = rate
            except HTTPException as exc:
                self.logger.warning(
                    "fx_bulk_rate_skipped",
                    cache_key=cache_key,
                    status_code=exc.status_code,
                    detail=exc.detail,
                )

        return results

    async def get_available_currencies(
        self,
        db: AsyncSession,
        as_of: date | None = None,
    ) -> list[str]:
        """Return sorted list of all unique currency codes in fx_rate.

        Always includes 'EUR'. Falls back to ['EUR'] on DB error.

        Args:
            db: Async DB session.
            as_of: Optional upper-bound date filter on rate_date.
        """
        try:
            from sqlalchemy import text as sa_text, union

            from_q = select(FxRate.from_ccy.label("ccy"))
            to_q = select(FxRate.to_ccy.label("ccy"))

            if as_of is not None:
                from_q = from_q.where(FxRate.rate_date <= as_of)
                to_q = to_q.where(FxRate.rate_date <= as_of)

            combined = union(from_q, to_q).subquery()
            stmt = select(combined.c.ccy).order_by(combined.c.ccy)
            rows = (await db.execute(stmt)).scalars().all()

            currencies: set[str] = set(rows)
            currencies.add("EUR")
            return sorted(currencies)
        except Exception as exc:
            self.logger.error("fx_get_currencies_failed", error=str(exc))
            return ["EUR"]

    # -----------------------------------------------------------------------
    # Private helpers
    # -----------------------------------------------------------------------

    async def _find_direct_rate(
        self,
        db: AsyncSession,
        from_ccy: str,
        to_ccy: str,
        rate_date: date,
    ) -> Decimal | None:
        stmt = (
            select(FxRate)
            .where(
                FxRate.from_ccy == from_ccy,
                FxRate.to_ccy == to_ccy,
                FxRate.rate_date <= rate_date,
            )
            .order_by(FxRate.rate_date.desc())
            .limit(1)
        )
        row = (await db.execute(stmt)).scalar_one_or_none()
        return Decimal(str(row.rate)) if row is not None else None

    async def _find_inverse_rate(
        self,
        db: AsyncSession,
        from_ccy: str,
        to_ccy: str,
        rate_date: date,
    ) -> Decimal | None:
        stmt = (
            select(FxRate)
            .where(
                FxRate.from_ccy == to_ccy,   # swapped
                FxRate.to_ccy == from_ccy,   # swapped
                FxRate.rate_date <= rate_date,
            )
            .order_by(FxRate.rate_date.desc())
            .limit(1)
        )
        row = (await db.execute(stmt)).scalar_one_or_none()
        return Decimal(str(row.rate)) if row is not None else None


# ---------------------------------------------------------------------------
# Singleton
# ---------------------------------------------------------------------------

_fx_service: FxService | None = None


def get_fx_service() -> FxService:
    global _fx_service
    if _fx_service is None:
        _fx_service = FxService()
    return _fx_service
